apps/builder/app/canvas/instance-selection.ts

```diff
@@ -0,0 +1,112 @@
+import { getComponentMeta, idAttribute } from "@webstudio-is/react-sdk";
+import {
+  instancesStore,
+  selectedInstanceAddressStore,
+  textEditingInstanceIdStore,
+} from "~/shared/nano-states";
+import { publish } from "~/shared/pubsub";
+import {
+  type InstanceAddress,
+  getAncestorInstanceAddress,
+} from "~/shared/tree-utils";
+
+declare module "~/shared/pubsub" {
+  export interface PubsubMap {
+    clickCanvas: undefined;
+  }
+}
+
+// traverse dom to the root and find all instances
+const getInstanceAddressFromElement = (element: Element) => {
+  const instanceAddress: InstanceAddress = [];
+  let matched: undefined | Element =
+    element.closest(`[${idAttribute}]`) ?? undefined;
+  while (matched) {
+    const instanceId = matched.getAttribute(idAttribute) ?? undefined;
+    if (instanceId !== undefined) {
+      instanceAddress.push(instanceId);
+    }
+    matched = matched.parentElement?.closest(`[${idAttribute}]`) ?? undefined;
+  }
+  if (instanceAddress.length === 0) {
+    return undefined;
```

Comment:
```suggestion
    return;
```

Reply:
also wouldn't it be better to just return empty array?

Reply:
the fact that one item always exists is something you would have to assume, in the downstream logic, its not guaranteed by types or checks